File: frontend/coordinator-android/engine/src/main/kotlin/ug/qaat/engine/AttendanceTrends.kt
package ug.qaat.engine

import kotlin.math.abs

/**
 * Attendance trend analytics (spec Feature 2) — pure logic over local data.
 * Given weekly (attended / possible) counts, computes the rate per week plus the
 * summary the coordinator dashboard shows. Computed on-device from SQLite.
 */
class AttendanceTrends(private val thresholdPct: Double = 75.0) {

    data class WeekInput(val label: String, val attended: Int, val possible: Int)
    data class WeekPoint(val label: String, val ratePct: Double, val belowThreshold: Boolean)
    enum class Trend { RISING, FALLING, STABLE }

    data class Summary(
        val weeks: List<WeekPoint>,
        val currentWeekRate: Double,
        val semesterAverage: Double,
        val bestWeek: WeekPoint?,
        val worstWeek: WeekPoint?,
        val totalSessions: Int,
        val weeksBelowThreshold: Int,
        val trend: Trend,
    )

    fun compute(weeks: List<WeekInput>): Summary {
        val points = weeks.map {
            val rate = if (it.possible > 0) it.attended.toDouble() / it.possible * 100.0 else 0.0
            WeekPoint(it.label, rate, rate < thresholdPct)
        }
        if (points.isEmpty())
            return Summary(emptyList(), 0.0, 0.0, null, null, 0, 0, Trend.STABLE)

        val current = points.last().ratePct
        // Semester average weighted by sessions (total attended / total possible).
        val totalAttended = weeks.sumOf { it.attended }
        val totalPossible = weeks.sumOf { it.possible }
        val avg = if (totalPossible > 0) totalAttended.toDouble() / totalPossible * 100.0 else 0.0
        val best = points.maxByOrNull { it.ratePct }
        val worst = points.minByOrNull { it.ratePct }
        val belowCount = points.count { it.belowThreshold }
        val trend = when {
            points.size < 2 -> Trend.STABLE
            else -> {
                val delta = points.last().ratePct - points[points.size - 2].ratePct
                if (abs(delta) < 1.0) Trend.STABLE else if (delta > 0) Trend.RISING else Trend.FALLING
            }
        }
        return Summary(points, current, avg, best, worst, weeks.size, belowCount, trend)
    }
}

File: frontend/coordinator-flutter/lib/engine/attendance_trends.dart
class WeekInput {
  final String label;
  final int attended;
  final int possible;

  const WeekInput(this.label, this.attended, this.possible);
}

class WeekPoint {
  final String label;
  final double ratePct;
  final bool belowThreshold;

  const WeekPoint(this.label, this.ratePct, this.belowThreshold);
}

enum Trend { rising, falling, stable }

class AttendanceSummary {
  final List<WeekPoint> weeks;
  final double currentWeekRate;
  final double semesterAverage;
  final WeekPoint? bestWeek;
  final WeekPoint? worstWeek;
  final int totalSessions;
  final int weeksBelowThreshold;
  final Trend trend;

  const AttendanceSummary(
    this.weeks,
    this.currentWeekRate,
    this.semesterAverage,
    this.bestWeek,
    this.worstWeek,
    this.totalSessions,
    this.weeksBelowThreshold,
    this.trend,
  );
}

class AttendanceTrends {
  final double thresholdPct;

  AttendanceTrends({this.thresholdPct = 75.0});

  AttendanceSummary compute(List<WeekInput> weeks) {
    final pts = [
      for (final w in weeks)
        WeekPoint(
          w.label,
          w.possible > 0 ? w.attended / w.possible * 100.0 : 0.0,
          w.possible > 0 ? w.attended / w.possible * 100.0 < thresholdPct : true,
        ),
    ];
    if (pts.isEmpty) {
      return const AttendanceSummary([], 0.0, 0.0, null, null, 0, 0, Trend.stable);
    }

    final current = pts.last.ratePct;
    final totalAttended = weeks.fold<int>(0, (acc, w) => acc + w.attended);
    final totalPossible = weeks.fold<int>(0, (acc, w) => acc + w.possible);
    final avg = totalPossible > 0 ? totalAttended / totalPossible * 100.0 : 0.0;

    WeekPoint? best;
    WeekPoint? worst;
    for (final p in pts) {
      if (best == null || p.ratePct > best.ratePct) best = p;
      if (worst == null || p.ratePct < worst.ratePct) worst = p;
    }
    final belowCount = pts.where((p) => p.belowThreshold).length;
    final Trend trend;
    if (pts.length < 2) {
      trend = Trend.stable;
    } else {
      final delta = pts.last.ratePct - pts[pts.length - 2].ratePct;
      if (delta.abs() < 1.0) {
        trend = Trend.stable;
      } else if (delta > 0) {
        trend = Trend.rising;
      } else {
        trend = Trend.falling;
      }
    }

    return AttendanceSummary(
      pts,
      current,
      avg,
      best,
      worst,
      weeks.length,
      belowCount,
      trend,
    );
  }
}